### ISSUE TITLE
Setting hyperparameter based on global step

### ISSUE THREAD

**Issue by MrWanter** [ppwwyyxx/tensorpack#633]
in `param.py`, we can only set hyperparams based on `epoch_num`:
```
class HyperParamSetterWithFunc(HyperParamSetter):
    """ Set the parameter by a function of epoch num and old value. """
    def __init__(self, param, func):
        """
        Args:
            param: same as in :class:`HyperParamSetter`.
            func: ``param`` will be set by ``new_value = func(epoch_num, old_value)``.
                ``epoch_num`` is the number of epochs that have finished.

        Example:
            Decrease by a factor of 0.9 every two epochs:

            .. code-block:: python

                HyperParamSetterWithFunc('learning_rate',
                                         lambda e, x: x * 0.9 if e % 2 == 0 else x)
        """
        super(HyperParamSetterWithFunc, self).__init__(param)
        self.f = func
```
but how can we set them by global steps, like tensorflow's example:
```
decayed_learning_rate = learning_rate *
                        decay_rate ^ (global_step / decay_steps)
```

**Comment by ppwwyyxx** [comment]
I plan to add such things to callbacks, but right now you can just write the above expression in the graph. You can get the global step with `tf.train.get_or_create_global_step()`.

**Comment by MrWanter** [comment]
I changed this two lines in param.py: https://github.com/ppwwyyxx/tensorpack/blob/bf4d89389b508844280d2bc6826441fe5dcd8243/tensorpack/callbacks/param.py#L281-L282

to    
```
def _get_value_to_set(self):
        return self.f(self.global_step, self.get_current_value())
```
then replace the original schedualer :
```
        ScheduledHyperParamSetter('learning_rate',
                                   [(0, 3e-1), (30, 3e-2), (60, 3e-3), (90, 3e-4)]),
```
with 
```
        HyperParamSetterWithFunc('learning_rate',
                                 lambda e, x: 0.5*(1-e/3e5)),
```
Am I right? but seems not right, after `10` epochs(global_step `12500`, total step `3e5`), lr still `0.4999`, not decrease to around `0.48`, by the way, seems the summary is also updated epoch based, can we modify it to `global_step` based?

**Comment by Skylion007** [comment]
So I have a really hacky way I like to do this using tf.cond() and variable.assign. Basically you can either use the default value or assign the value based on whether the condition is true. This is what I have done whenever i need to decay the learning rate.

**Comment by ppwwyyxx** [comment]
As I said, the easiest way to do it now is to do it in the graph, with `tf.train.get_or_create_global_step()`. You can also write a new callback to do this. Hacking with the existing one will not work.

>  the summary is also updated epoch based, can we modify it to global_step based?

It can and is mentioned in docs. http://tensorpack.readthedocs.io/en/latest/tutorial/summary.html

**Comment by MrWanter** [comment]
I change the line:
https://github.com/ppwwyyxx/tensorpack/blob/bf4d89389b508844280d2bc6826441fe5dcd8243/tensorpack/callbacks/param.py#L161
to  `def _trigger_step(self):` and it works properly!

**Comment by MrWanter** [comment]
@Skylion007 Thanks for sharing! I'll try it

**Comment by MrWanter** [comment]
as in the http://tensorpack.readthedocs.io/en/latest/tutorial/summary.html :
> All the "what, when, where" can be customized in either the graph or with the callbacks/monitors setting.

say,  I only want to change the lr to be step based, keep other summary epoch based, 
for
> with the callbacks/monitors setting 

seems tensorpack put all summaries writing together:
https://github.com/ppwwyyxx/tensorpack/blob/bf4d89389b508844280d2bc6826441fe5dcd8243/tensorpack/callbacks/monitor.py#L120-L141

is it difficult to modify the code since it is wrapped deeply?

for
> can be customized in either the graph

I dont understand what you mean by customize it graph, do you mean call `tf.summary.xxx` when we construct the graph? 

can you share a simple example of using `tf.train.get_or_create_global_step()`?

**Comment by ppwwyyxx** [comment]
What you asked for in the original issue is just one line of code in the graph.
https://github.com/ppwwyyxx/tensorpack/blob/9bbdf94da8be9976725fb131bfb5c9f5523b25ac/examples/mnist-convnet.py#L86-L90

**Comment by ppwwyyxx** [comment]
```diff
diff --git i/examples/mnist-convnet.py w/examples/mnist-convnet.py
index 9459533..c94ed57 100755
--- i/examples/mnist-convnet.py
+++ w/examples/mnist-convnet.py
@@ -90,7 +90,7 @@ class Model(ModelDesc):
             decay_rate=0.3, staircase=True, name='learning_rate')
         # This will also put the summary in tensorboard, stat.json and print in terminal
         # but this time without moving average
-        tf.summary.scalar('lr', lr)
+        tf.summary.scalar('lr', lr, collections=['NEW_COLL'])
         return tf.train.AdamOptimizer(lr)
 
 
@@ -116,6 +116,10 @@ def get_config():
             InferenceRunner(    # run inference(for validation) after every epoch
                 dataset_test,   # the DataFlow instance used for validation
                 ScalarStats(['cross_entropy_loss', 'accuracy'])),
+            MergeAllSummaries(period=1, key='NEW_COLL')
+        ],
+        monitors=DEFAULT_MONITORS() + [
+            ScalarPrinter(enable_step=True, enable_epoch=False)
         ],
         steps_per_epoch=steps_per_epoch,
         max_epoch=100,
```
This makes mnist example print learning rate every step.

**Comment by MrWanter** [comment]
Thank you, a great example for customizing!

**Comment by MrWanter** [comment]
and for `tf.train.get_or_create_global_step`, do you mean write a `lr` decay function like `tf.train.exponential_decay` and pass it to optimizer?

**Comment by ppwwyyxx** [comment]
Yes. 
tf.train.piecewise_constant, tf.train.exponential_decay can be useful but you can just write any symbolic function to compute your lr.

**Comment by MrWanter** [comment]
I see, thank you
